### PR TITLE
drivers: ieee802154: nrf5: multiple CCA support for OT

### PR DESCRIPTION
**Pull request by ankuns**
This PR is an alternative to https://github.com/zephyrproject-rtos/zephyr/pull/60343

It reverts changes made by https://github.com/zephyrproject-rtos/zephyr/pull/60390 (justification https://github.com/zephyrproject-rtos/zephyr/pull/60390#issuecomment-1636129759)

For nrf802154_radio.h extension points are added that allow to extend enums by higher layer.

OpenThread-specific capability "Multiple CCA" is added as a separate file, leveraging exposed extension points provided by nrf820154_radio.h

Implementation of the "Multiple CCA" capability is added to the ieee802154_nrf5 driver.

### CODE REVIEW

**Comment by ankuns** [comment]
@jciupis @ahasztag @LuDuda @edmont Please review and let me know if this satisfies your needs

File: include/zephyr/net/ieee802154_radio.h
Comment (jciupis) [inline include/zephyr/net/ieee802154_radio.h:118]: `Note`?

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (jciupis) [inline include/zephyr/net/ieee802154_radio_ext_ot.h:28]: I would add the information that the CCA attempts are performed back to back and that the transmission starts as soon as idle channel is detected. The following sentence from `ieee802154_ext_ot_config` documentation can be copied here:
> The CCA attempts are performed back-to-back until CCA returns idle channel or until 1 + extra_cca_attempts CCA attempts are performed.

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (jciupis) [inline drivers/ieee802154/ieee802154_nrf5.c:234]: I'm wondering if we need this symbol at all. The nRF driver can support it unconditionally, the memory & processing overheads seem negligible. What do you think?

Comment (ahasztag) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: I think that as this is not a standard, specification compliant feature it should be optional.

Comment (ankuns) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: The nRF driver needs to be deployed first, won't compile otherwise.

**Review by ahasztag — CHANGES_REQUESTED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (ahasztag) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: I think that as this is not a standard, specification compliant feature it should be optional.

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (ankuns) [reply drivers/ieee802154/ieee802154_nrf5.c:1169]: Please follow commit-by commit, First the commit was reverted, but later some fields are added (with the same names) because they are useful.

File: include/zephyr/net/ieee802154_radio.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio.h]: fixed

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (jciupis) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: > The nRF driver needs to be deployed first, won't compile otherwise.

True, however it's a matter of several days. If that's the only reason for introducing this switch then in my opinion we can wait and remove it.

> I think that as this is not a standard, specification compliant feature it should be optional.

It's already supported unconditionally by the driver, the capability is controlled by ieee802154 layer. What's the benefit from the possibility to disable it here?

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (ahasztag) [reply drivers/ieee802154/ieee802154_nrf5.c:1169]: I think my comment is still valid - `nrf5_radio->extra_cca_attempts` is used in `nrf5_tx_at` yet there does not seem to be a place where it is assigned to...

Comment (jciupis) [reply drivers/ieee802154/ieee802154_nrf5.c:1169]: It's assigned in `nrf5_configure` :)

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (ankuns) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: If you don't use this feature you can keep turned it off, this saves ram&flash. Linker won't remove this code because implementation is referenced through function pointers. 
Even if you never call 
`radio_api->configure` with `IEEE802154_EXT_OT_CONFIG_MULTIPLE_CCA` or 
`radio_api->tx` with mode == IEEE802154_EXT_OT_TX_MODE_TXTIME_MULTIPLE_CCA
the code handling this cases will need to remain there.

Comment (jciupis) [reply drivers/ieee802154/ieee802154_nrf5.c:234]: I agree with you. My point was that these savings are negligible and not worth the additional configuration, especially that the driver below supports the feature unconditionally anyway, so disabling the option here doesn't lead to any more savings there.

I'm not blocking the PR over it ofc.

**Review by jciupis — DISMISSED** [review]

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (ahasztag) [inline drivers/ieee802154/ieee802154_nrf5.c:568]: Actually, is there really need to make a separate case for `IEEE802154_EXT_OT_TX_MODE_TXTIME_MULTIPLE_CCA` ?  If `extra_cca_attempts` is zero this will have exactly the same functionality as `IEEE802154_TX_MODE_TXTIME_CCA`.

I don't see really a reason for keeping it as a separate case.

Comment (jciupis) [reply drivers/ieee802154/ieee802154_nrf5.c:568]: It could be handled as you described. However, in my opinion the current implementation provides better separation of concepts, which was the main motivation behind replacing #60343 with this PR:
* `IEEE802154_TX_MODE_TXTIME_CCA` is defined in common namespace as it's a well-defined concept useful in scope of IEEE 802.15.4 spec
* `IEEE802154_EXT_OT_TX_MODE_TXTIME_MULTIPLE_CCA` is an OpenThread-specific extension that is not a part of the common IEEE 802.15.4 interface

I think it's cleaner than handling both behaviors using `IEEE802154_TX_MODE_TXTIME_CCA` since clear separation of common 802.15.4 API and custom extensions is essentially the reason behind this PR

**Review by rlubos — DISMISSED** [review]
LGTM, from my perspective it looks exactly as I imagined.

**Review by ahasztag — APPROVED** [review]

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (edmont) [inline include/zephyr/net/ieee802154_radio_ext_ot.h:57]: Clarify between two options? I assume it's first one, but worth making it clear.

1. when transmission is requested to the radio driver by the upper layer
2. when transmission is requested to the lower layer by the radio driver (after the desired delay)

**Comment by fgrandel** [comment]
Just added DNM so I have time to review this (will be done now) and we don't accidentally merge in the middle again. :-)

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (edmont) [reply drivers/ieee802154/ieee802154_nrf5.c:568]: Both options would make possible to perform per frame selection of the CCA attempts number just by extending the transmission information metadata of the `otRadioFrame`. Once the metadata needs to be updated anyway, it seems more efficient for me to be able to configure the number of extra CCA attempts rather than single/multiple attempts plus a separated radio API for configuring the number of attempts.

So basically no big preference for any of the two options. But I would tend to go for having `IEEE802154_EXT_OT_TX_MODE_TXTIME_MULTIPLE_CCA` for the reasons @jciupis described.

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (fgrandel) [reply include/zephyr/net/ieee802154_radio_ext_ot.h:57]: I agree that the timing requirements should be documented (as we agreed this shows the way and is a precondition for a potential future soft MAC solution w/o the need to implement it right away). The easiest way would be to just point to the (future) standards clause but that does not seem to be possible, so alternatively the actual requirements need to be documented IMO:

1) For the other timed TX modes, the TX timestamp points to the expected start of the first symbol of the PHY header (i.e. after CCA completed) for packets scheduled to a future point in time. What does the TX timestamp point to here? The first symbol of the PHY header assuming the first CCA attempt shows a free medium? The second CCA attempt if the first was showing a busy medium? The first CCA attempt?

2) It has been variously stated that a cross-vendor soft MAC approach was not possible due to strict timing requirements for the extra CCA attempts. What are these strict requirements? What is the duration of each CCA attempt? How are the CCA attempts scheduled? Or alternatively: What are the requirements re guard times between the attempts? Which tolerances (drift, jitter) are acceptable?

3) How is CCA performed (ie. which of the standard CCA modes are allowed here)?

Docs tend to become outdated and redundant docs even more quickly: I therefore propose that we keep the explanation as precise as possible but only in the TX_MODES enum and @ref to it here.

File: include/zephyr/net/ieee802154_radio.h
Comment (fgrandel) [inline include/zephyr/net/ieee802154_radio.h:127]: brackets not required here

File: include/zephyr/net/ieee802154_radio.h
Comment (fgrandel) [inline include/zephyr/net/ieee802154_radio.h:315]: config values -> types (see the API's nomenclature)

**Review by ghost — CHANGES_REQUESTED** [review]
I'm obsessing over names and docs here because they are conceptually important (apart from documenting the API to community developers of course): Semantically correct names and docs show full understanding and agreement about the underlying architecture. Plus they are important for enabling correct alternative soft-MAC implementations in the future.

We're introducing a new architectural pattern here that is meant to be extensible so I think it's worth the effort to try to get it right plus I'm always available for f2f pairing if that helps.

Nothing fundamental, though, this now feels like it is just the right solution for the problem. :-)

File: drivers/ieee802154/ieee802154_nrf5.c
Comment (fgrandel) [reply drivers/ieee802154/ieee802154_nrf5.c:568]: > I think it's cleaner than handling both behaviors using `IEEE802154_TX_MODE_TXTIME_CCA` since clear separation of common 802.15.4 API and custom extensions is essentially the reason behind this PR

+1

@ahasztag In addition to @ankuns striking argument re supporting alternating calls: Having one option only will make the call site conceptually ambiguous (and thereby hard to maintain/read). It's important that we cleanly separate calls that implement IEEE 802.15.4 standard primitives (and document them as such by pointing explicitly to the spec) from calls that implement Thread standard primitives (and document them as well as such) *at the call site*. See https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1220059230 for a detailed discussion of the rationale behind that approach - TLDR; it's to ensure cross-L2 re-usability - i.e. 15.4 standard primitives will also be used by native L2 while Thread primitives will not. Native L2 expects standard behavior for all calls no matter what the current configuration.

@ankuns I think we need to be careful to always think "API first" (aka *standards first*). We're not implementing low level features and then wrap some API around it. We're developing an API that makes sense from a conceptual/architectural standpoint because it is based (and points to) standard primitives and THEN we provide a low-level driver implementation for it. I mention this explicitly to everyone here without meaning you personally as several recent API additions clearly seem to have originated as simple wrappers around nRF driver features which is what brought us in a difficult situation (again see the link above). So maybe in the future API changes should be discussed in RFCs with separate PRs before an implementation in the nRF driver is attempted - this will considerably reduce the risk of blocked PRs (and thereby actual review/implementation effort) and is a very common approach for API changes in Zephyr AFAIK.

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: fixed

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: changed to `max_extra_cca_attempts` along with config type to `IEEE802154_OPENTHREAD_CONFIG_MAX_EXTRA_CCA_ATTEMPTS`

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: I believe I switched to simple "CCA" or "CCAs" everywhere, please have a look

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (rlubos) [reply include/zephyr/net/ieee802154_radio_ext_ot.h:1]: > L2_OPENTHREAD is enabling a specific implementation of the standard, not the standard itself - so I think this supports my argument (that was exactly the fine line I was trying to draw).

But that's my point - we're not implementing Thread standard in Zephyr, we're integrating 3rd party Thread implementation (OpenThread), therefore I'd prefer to stick to a uniform naming convention for the related code.

> We could have L2_XXXTHREAD at any time. But then L2_OPENTHREAD and L2_XXXTHREAD would share the same drivers and therefore the same driver options as the feature is common to the Thread standard. It would be strange if L2_XXXTHREAD would have to call configure(..., IEEE802154_OPENTHREAD_...).

That's a rather unspecified future. And if we (ever) come up to a point where we want to support another Thread implementation, we'd likely need to make more architectural choices in terms of generalizing Thread support in Zephyr (like for example for me personally, it'd then make sense then to switch to a common `L2_THREAD` with underlying implementation choice instead of multiplying L2s).

> Would you leave out the "EXT"? This would then result in:
> 
> file: ieee802154_radio_ot.h (as before) prefix: IEEE802154_OT_{CONFIG|HW|TX_MODE}_...
> 
> Or would you prefer:
> 
> file: ieee802154_radio_openthread.h prefix: IEEE802154_OPENTHREAD_{CONFIG|HW|TX_MODE}_...

We can leave out "EXT". For me personally, in 15.4 context, "OT" is just as fine as full "OPENTHREAD", but it might not be the case for other, less familiar with OpenThread, so perhaps in terms of clarity, "OPENTHREAD" could be a better choice here.

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: `IEEE802154_OPENTHREAD_CONFIG_MAX_EXTRA_CCA_ATTEMPTS` used
`OPENTHREAD` instead of `THREAD` picked, justification shortly in replies to https://github.com/zephyrproject-rtos/zephyr/pull/60517#discussion_r1266829701

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: Stayed at `OpenThread`, will justify in discussion following https://github.com/zephyrproject-rtos/zephyr/pull/60517#discussion_r1266829701

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h:1]: Picked scheme:

file: ieee802154_radio_openthread.h
prefix: IEEE802154_OPENTHREAD_{CONFIG|HW|TX_MODE}_...

`ext`/`EXT` infix removed

File: include/zephyr/net/ieee802154_radio.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio.h]: fixed

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: renamed to `common`

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: I suppose It's fixed, used scheme as for ieee802154_radio.h, inserted `@brief`

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: Fixed, added one sentence and `@ref`

File: include/zephyr/net/ieee802154_radio.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio.h]: fixed, removed brackets

File: include/zephyr/net/ieee802154_radio.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio.h]: fixed

**Comment by fgrandel** [comment]
@ankuns Thanks for the fixes! All resolved except for the timing spec according to https://github.com/zephyrproject-rtos/zephyr/pull/58438#discussion_r1267034840. Let me know if I can help in formulating that - as I said I'm always available for pairing and/or on Discord to keep the effort minimal on your side.

**Comment by fgrandel** [comment]
@ankuns The failing build check is due to #60551 - so nothing you can do about it. Just rebase once that's merged.

**Comment by ankuns** [comment]
just rebased, will amend content in the next push

File: include/zephyr/net/ieee802154_radio_ext_ot.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_ext_ot.h]: I agree, that needs to be addressed. I will return with proposal.

**Review by jciupis — DISMISSED** [review]
Flawless. Kudos to you for the documentation 🚀

**Review by ahasztag — APPROVED** [review]
Looks good to me

**Review by LuDuda — APPROVED** [review]

File: include/zephyr/net/ieee802154_radio_openthread.h
Comment (edmont) [inline include/zephyr/net/ieee802154_radio_openthread.h:35]: This line is misleading. As the doc for the file specifies the features added here are to be part of OpenThread, but that doesn't imply that they must be part of Thread Specification. This extension should allow to add any proprietary features on top of Thread Specification minimum requirements.

I'd suggest to just remove this line. The feature itself is quite well explained here so there is no need to refer to any external specification.

**Review by edmont — CHANGES_REQUESTED** [review]
Looks great, just a comment regarding documentation.

File: include/zephyr/net/ieee802154_radio_openthread.h
Comment (fgrandel) [reply include/zephyr/net/ieee802154_radio_openthread.h:35]: @edmont Robert had similar feedback. He proposed on Discord that we could simply call this an "experimental OpenThread feature".

So I'd propose wording similar to:

"This mode is a non-standard experimental OpenThread feature. It allows transmission of a packet within a certain time window."

**Review by rlubos — DISMISSED** [review]

**Review by ghost — DISMISSED** [review]

File: include/zephyr/net/ieee802154_radio_openthread.h
Comment (ankuns) [reply include/zephyr/net/ieee802154_radio_openthread.h]: Changed the sentence to Florian's proposal.

**Comment by ankuns** [comment]
Addressed all issues, waiting for hopefully final approvals @fgrandel @rlubos @jciupis

**Review by edmont — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by jciupis — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**
